test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +53,475 @@ var (
-		for _, p := range test.InitPods {
-			initPods += p.Num
-		}
-		name := fmt.Sprintf("%v/%vNodes/%vInitPods/%vPodsToSchedule", test.Desc, test.Nodes.Num, initPods, test.PodsToSchedule.Num)
-		b.Run(name, func(b *testing.B) {
-			for feature, flag := range test.FeatureGates {
-				defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, feature, flag)()
+	dataItems := DataItems{Version: "v1"}
+	for _, tc := range testCases {
+		b.Run(tc.Name, func(b *testing.B) {
+			for _, w := range tc.Workloads {
+				b.Run(w.Name, func(b *testing.B) {
+					for feature, flag := range tc.FeatureGates {
+						resetFunc := featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, feature, flag)
+						defer resetFunc()
+					}
+					dataItems.DataItems = append(dataItems.DataItems, runWorkload(b, tc, w)...)
+				})
 			}
-			dataItems.DataItems = append(dataItems.DataItems, perfScheduling(test, b)...)
 		})
 	}
 	if err := dataItems2JSONFile(dataItems, b.Name()); err != nil {
 		klog.Fatalf("%v: unable to write measured data: %v", b.Name(), err)
 	}
 }
 
-func perfScheduling(test testCase, b *testing.B) []DataItem {
+func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+	var mu sync.Mutex
+	var dataItems []DataItem
 	finalFunc, podInformer, clientset := mustSetupScheduler()
 	defer finalFunc()
 
-	nodePreparer, err := getNodePreparer(test.Nodes, clientset)
-	if err != nil {
-		b.Fatal(err)
-	}
-	if err := nodePreparer.PrepareNodes(); err != nil {
-		b.Fatal(err)
-	}
-	defer nodePreparer.CleanupNodes()
-
-	total := 0
-	for _, p := range test.InitPods {
-		if err := createPods(setupNamespace, p, clientset); err != nil {
-			b.Fatal(err)
-		}
-		total += p.Num
-	}
-	if !test.SkipWaitUntilInitPodsScheduled {
-		if err := waitNumPodsScheduled(b, total, podInformer, setupNamespace); err != nil {
-			b.Fatal(err)
-		}
-	}
+	numPodsScheduledPerNamespace := make(map[string]int)
+	numNodes := 0
+	allPodsAsync := true
 
-	// start benchmark
+	b.StopTimer()
 	b.ResetTimer()
+	for opIndex, op := range w.Ops {
+		switch realOp := op.realOp.(type) {
+		case *createNodesOp:
+			nodePreparer, err := getNodePreparer(fmt.Sprintf("node-%d-", opIndex), realOp, clientset)
+			if err != nil {
+				b.Fatal(err)
+			}
+			if err := nodePreparer.PrepareNodes(numNodes); err != nil {
+				b.Fatal(err)
+			}
+			if numNodes == 0 {
+				// Schedule a cleanup at most once. The CleanupNodes function will list
+				// and delete *all* nodes.
+				// TODO(adtac): make CleanupNodes only clean up its own nodes to make
+				// this more intuitive?
+				defer nodePreparer.CleanupNodes()
+			}
+			numNodes += realOp.CreateNodes
+
+		case *createPodsOp:
+			var namespace string
+			if realOp.Namespace != nil {
+				namespace = *realOp.Namespace
+			} else {
+				namespace = fmt.Sprintf("namespace-%d", opIndex)
+			}
+			var stopCh chan struct{}
+			var collectors []testDataCollector
+			if realOp.CollectMetrics {
+				b.StartTimer()
```

Comment:
I don't think we should care too much about when to start/end this timer, we should just make it span the whole benchmark case, from beginning to end. This is because we will look mostly at the metrics (throughput, latency etc.).

Reply:
Ack, done.

Reply:
StartTimer call still exist, I am suggesting to remove this.